70_FLIPFluids/app_resources/compute/updateFluidCells.comp.hlsl

```diff
@@ -0,0 +1,120 @@
+
+    uint thisCellMaterial = getCellMaterial(cellMaterialInBuffer[tid]);
+    uint cellMaterial = 0;
+    setCellMaterial(cellMaterial, thisCellMaterial);
+
+    uint xpCm = cIdx.x == 0 ? CM_SOLID : getCellMaterial(cellMaterialInBuffer[cellIdxToFlatIdx(cIdx + int3(-1, 0, 0), gridData.gridSize)]);
+    setXPrevMaterial(cellMaterial, xpCm);
+
+    uint xnCm = cIdx.x == gridData.gridSize.x - 1 ? CM_SOLID : getCellMaterial(cellMaterialInBuffer[cellIdxToFlatIdx(cIdx + int3(1, 0, 0), gridData.gridSize)]);
+    setXNextMaterial(cellMaterial, xnCm);
+
+    uint ypCm = cIdx.y == 0 ? CM_SOLID : getCellMaterial(cellMaterialInBuffer[cellIdxToFlatIdx(cIdx + int3(0, -1, 0), gridData.gridSize)]);
+    setYPrevMaterial(cellMaterial, ypCm);
+
+    uint ynCm = cIdx.y == gridData.gridSize.y - 1 ? CM_SOLID : getCellMaterial(cellMaterialInBuffer[cellIdxToFlatIdx(cIdx + int3(0, 1, 0), gridData.gridSize)]);
+    setYNextMaterial(cellMaterial, ynCm);
+
+    uint zpCm = cIdx.z == 0 ? CM_SOLID : getCellMaterial(cellMaterialInBuffer[cellIdxToFlatIdx(cIdx + int3(0, 0, -1), gridData.gridSize)]);
+    setZPrevMaterial(cellMaterial, zpCm);
+
+    uint znCm = cIdx.z == gridData.gridSize.z - 1 ? CM_SOLID : getCellMaterial(cellMaterialInBuffer[cellIdxToFlatIdx(cIdx + int3(0, 0, 1), gridData.gridSize)]);
+    setZNextMaterial(cellMaterial, znCm);
+
+    cellMaterialOutBuffer[tid] = cellMaterial;
+}
+
+[numthreads(WorkgroupSize, 1, 1)]
+void addParticlesToCells(uint32_t3 ID : SV_DispatchThreadID)
+{
+    uint tid = ID.x;
+    int3 cIdx = flatIdxToCellIdx(tid, gridData.gridSize);
+
+    float3 position = cellIdxToWorldPos(cIdx, gridData);
+    float3 posvx = position + float3(-0.5f * gridData.gridCellSize, 0.0f, 0.0f);
+    float3 posvy = position + float3(0.0f, -0.5f * gridData.gridCellSize, 0.0f);
+    float3 posvz = position + float3(0.0f, 0.0f, -0.5f * gridData.gridCellSize);
+
+    float3 totalWeight = 0;
+    float3 totalVel = 0;
+
+    for (int i = max(cIdx.x - 1, 0); i <= min(cIdx.x + 1, gridData.gridSize.x - 1); i++)
+    {
+        for (int j = max(cIdx.y - 1, 0); j <= min(cIdx.y + 1, gridData.gridSize.y - 1); j++)
+        {
+            for (int k = max(cIdx.z - 1, 0); k <= min(cIdx.z + 1, gridData.gridSize.z - 1); k++)
+            {
+                uint2 idx = gridParticleIDBuffer[cellIdxToFlatIdx(int3(i, j, k), gridData.gridSize)];
+                for (uint pid = idx.x; pid < idx.y; pid++)
+                {
+                    Particle p = particleBuffer[pid];
+                    
+                    float3 weight;
+                    weight.x = getWeight(p.position.xyz, posvx, gridData.gridInvCellSize);
+                    weight.y = getWeight(p.position.xyz, posvy, gridData.gridInvCellSize);
+                    weight.z = getWeight(p.position.xyz, posvz, gridData.gridInvCellSize);
+
+                    totalWeight += weight;
+                    totalVel += weight * p.velocity.xyz;
+                }
```

Comment:
this has somewhat bad load balancing, there is another way to do this, ATOMICS

Basically you'd need to clear the velocity buffer to 0 in some previous dispatch and in `updateFluidCells` (since you're going through particles anyway) you'd do atomicAdd of the weighted velocity to the cell.

why this works?
- you're supposed to split the velocity buffer into mono-channel anyway (3 images one per component)
- emulated float atomic ops can be done via CAS-loops on R32_UINT underlying
- there are also extensions for atomic float and atomic float16 (would have to check if we list them in SFeatures or SLimits of the physical device) if you feel like specializing with device traits

E.g. CAS loop atomic
```cpp
float32_t oldValue = 0.f;
do
{
   float32_t expectedValue = oldValue;
   float32_t newValue = oldValue+weightedVelocity;
   uint32_t oldBitpattern;
   InterlockedCompareExchange(velocity[component][cIdx],asuint(expectedValue),asuint(newValue),oldBitpattern);
   oldValue = asfloat(oldBitpattern);
} while (oldValue!=expectedValue);
```

Reply:
then if all the weighted accumulation happensin `updateFluidCells` this dispatch only needs to enforce the boundary condition so can be kernel-fusioned with the `updateNeighbour` dispatch

Reply:
Yep there should be `device_traits` for `shaderBufferFloat16AtomicAdd`, `shaderBufferFloat32AtomicAdd`, `shaderImageFloat32AtomicAdd` if you compile your code with `ILogicalDevice`